community/primitive-collections/src/main/java/org/neo4j/concurrent/WorkSync.java

```diff
@@ -0,0 +1,181 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.concurrent;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * Turns multi-threaded unary work into single-threaded stack work.
+ * <p>
+ *     The technique used here is inspired in part both by the Flat Combining
+ *     concept from Hendler, Incze, Shavit &amp; Tzafrir, and in part by the
+ *     wait-free linked queue design by Vyukov.
+ * </p>
+ * <p>
+ *     In a sense, this turns many small, presumably concurrent, pieces of work
+ *     into fewer, larger batches of work, that is then applied to the material
+ *     under synchronisation.
+ * </p>
+ * <p>
+ *     Obviously this only makes sense for that a) can be combined, and b)
+ *     where the performance improvements from batching effects is large enough
+ *     to overcome the overhead of collecting and batching up the work units.
+ * </p>
+ * @see Work
+ */
+public class WorkSync<Material, W extends Work<Material,W>>
+{
+    private final Material material;
+    private final AtomicReference<WorkUnit<Material,W>> stack;
+    private final WorkUnit<Material,W> stackEnd;
+    private final ReentrantLock lock;
+
+    /**
+     * Create a new WorkSync that will synchronize the application of work to
+     * the given material.
+     * @param material The material we want to apply work to, in a thread-safe
+     * way.
+     */
+    public WorkSync( Material material )
+    {
+        this.material = material;
+        this.stackEnd = new WorkUnit<>( null );
+        this.stack = new AtomicReference<>( stackEnd );
+        this.lock = new ReentrantLock();
+    }
+
+    /**
+     * Apply the given work to the material in a thread-safe way, possibly by
+     * combining it with other work.
+     * @param work The work to be done.
+     */
+    public void apply( W work )
+    {
+        // Schedule our work on the stack.
+        WorkUnit<Material,W> unit = new WorkUnit<>( work );
+        unit.next = stack.getAndSet( unit ); // benign race, see reverse()
+
+        // Try grabbing the lock to do all the work, until our work unit
+        // has been completed.
+        boolean wasInterrupted = false;
+        int tryCount = 0;
+        do
+        {
+            tryCount++;
+            try
+            {
+                if ( lock.tryLock( tryCount < 10? 0 : 10, TimeUnit.MILLISECONDS ) )
+                {
+                    doSynchronizedWork();
```

Comment:
An exception thrown here would not call `lock.unlock()` below. Shouldn't this be:

```
try
{
    doSynchronizedWork();
}
finally
{
    lock.unlock();
}
```

instead?